iOS/Collection View Management /CollectionViewDataSource.swift

```diff
@@ -318,6 +318,28 @@ extension CollectionViewDataSource {
 // Functions to do custom animations to the collection view in conjunctions with applying snapshots.
 extension CollectionViewDataSource {
 
+    /// Applies the snapshot while adjusting the related collectionview's content offset so that the visible cells don't move.
+    /// For example, say there are 4 cells and cells 1 and 2 are visible:  0 [1  2] 3 (brackets symbolize viewport)
+    /// If 2 more items are are added to the beginning of the collectionview, the viewport will stay centered on the items like so:
+    /// -2 -1 0 [1 2] 3
+    @MainActor
+    func applySnapshotKeepingVisualOffset(_ snapshot: SnapshotType, collectionView: UICollectionView) async {
+        // Note the content size before any updates are applied
+        let beforeContentSize = collectionView.contentSize
+
+        // Reload
+        await self.apply(snapshot, animatingDifferences: false)
+
+        collectionView.layoutIfNeeded()
+        let afterContentSize = collectionView.contentSize
+
+        // reset the contentOffset after data is updated
+        let newOffset = CGPoint(
+            x: collectionView.contentOffset.x + (afterContentSize.width - beforeContentSize.width),
+            y: collectionView.contentOffset.y + (afterContentSize.height - beforeContentSize.height))
+        collectionView.setContentOffset(newOffset, animated: false)
+    }
+
```

Comment:
Basically the same thing as the reload and keep offset except we apply a snapshot to a data source.

Reply:
nice